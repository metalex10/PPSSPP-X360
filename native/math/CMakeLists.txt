set(SRCS
  lin/matrix4x4.cpp
  lin/vec3.cpp
  lin/quat.cpp
  lin/aabb.cpp
	curves.cpp
	math_util.cpp
  )

set(SRCS ${SRCS})

add_library(lin STATIC ${SRCS})

if(UNIX)
  add_definitions(-fPIC)
endif(UNIX)
