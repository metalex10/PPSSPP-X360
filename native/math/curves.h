#pragma once

#include <base/basictypes.h>

// Easy curve computation for fades etc.


// output range: [0.0, 1.0]
float linearInOut(int t, int fadeInLength, int solidLength, int fadeOutLength);
float linearIn(int t, int fadeInLength);
float linearOut(int t, int fadeInLength);

// smooth operator [0, 1] -> [0, 1]
float ease(float val);
float ease(int t, int fadeLength);

// need a bouncy ease

// waveforms [0, 1]
float sawtooth(int t, int period);

// output range: -1.0 to 1.0
float passWithPause(int t, int fadeInLength, int pauseLength, int fadeOutLength);
