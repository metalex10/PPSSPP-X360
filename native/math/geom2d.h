#pragma once

#include <cmath>

struct Point {
	Point() {}
	Point(float x_, float y_) : x(x_), y(y_) {}

	float x;
	float y;

	float distanceTo(const Point &other) const {
		float dx = other.x - x, dy = other.y - y;
		return sqrtf(dx*dx + dy*dy);
	}

	/*
	FocusDirection directionTo(const Point &other) const {
		int angle = atan2f(other.y - y, other.x - x) / (2 * M_PI) - 0.125;

	}*/
};


// Resolved bounds on screen after layout.
struct Bounds {
	Bounds() : x(0), y(0), w(0), h(0) {}
	Bounds(float x_, float y_, float w_, float h_) : x(x_), y(y_), w(w_), h(h_) {}

	bool Contains(float px, float py) const {
		return (px >= x && py >= y && px < x + w && py < y + h);
	}

	bool Intersects(const Bounds &other) const {
		return !(x > other.x2() || x2() < other.x || y > other.y2() || y2() < other.y);
	}

	void Clip(const Bounds &clipTo) {
		if (x < clipTo.x) {
			w -= clipTo.x - x;
			x = clipTo.x;
		}
		if (y < clipTo.y) {
			h -= clipTo.y - y;
			y = clipTo.y;
		}
		if (x2() > clipTo.x2()) {
			w = clipTo.x2() - x;
		}
		if (y2() > clipTo.y2()) {
			h = clipTo.y2() - y;
		}
	}

	float x2() const { return x + w; }
	float y2() const { return y + h; }
	float centerX() const { return x + w * 0.5f; }
	float centerY() const { return y + h * 0.5f; }
	Point Center() const {
		return Point(centerX(), centerY());
	}

	float x;
	float y;
	float w;
	float h;
};


