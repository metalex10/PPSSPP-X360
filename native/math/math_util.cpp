#include "math/math_util.h"
#include <stdlib.h>

#if defined(__ARM_ARCH_7A__) && !defined(BLACKBERRY)

void EnableFZ()
{
	int x;
	asm(
		"fmrx %[result],FPSCR \r\n"
		"orr %[result],%[result],#16777216 \r\n"
		"fmxr FPSCR,%[result]"
		:[result] "=r" (x) : :
	);
	//printf("ARM FPSCR: %08x\n",x);
}
#else
void EnableFZ()
{
	// TODO
}
#endif
