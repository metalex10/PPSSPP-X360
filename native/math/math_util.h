#ifndef _MATHUTILS_H
#define _MATHUTILS_H

#include <cmath>
#include <cstring>

#include "base/basictypes.h"

inline float sqr(float f)	{return f*f;}
inline float sqr_signed(float f) {return f<0 ? -f*f : f*f;}

typedef unsigned short float16;

// This ain't a 1.5.10 float16, it's a stupid hack format where we chop 16 bits off a float.
// This choice is subject to change. Don't think I'm using this for anything at all now anyway.
// DEPRECATED
inline float16 FloatToFloat16(float x) {
	int ix;
	memcpy(&ix, &x, sizeof(float));
	return ix >> 16;
}

inline float Float16ToFloat(float16 ix) {
	float x;
	memcpy(&x, &ix, sizeof(float));
	return x;
}

inline uint32_t RoundUpToPowerOf2(uint32_t v) {
	v--;
	v |= v >> 1;
	v |= v >> 2;
	v |= v >> 4;
	v |= v >> 8;
	v |= v >> 16;
	v++;
	return v;
}

#define PI 3.141592653589793f
#ifndef M_PI
#define M_PI 3.141592653589793f
#endif

// The stuff in this file is from all over the web, esp. dspmusic.org. I think it's all public domain.
// In any case, very little of it is used anywhere at the moment.

// Calculate pseudo-random 32 bit number based on linear congruential method. 
void SetSeed(unsigned int seed);
unsigned int GenerateRandomNumber();
inline float GenerateRandomFloat01() {
	return (float)((double)GenerateRandomNumber() / 0xFFFFFFFF);
}
inline float GenerateRandomSignedFloat() {
	return (float)((double)GenerateRandomNumber() / 0x80000000) - 1.0f;
}


inline float GaussRand()
{
	float R1 = GenerateRandomFloat01();
	float R2 = GenerateRandomFloat01();

	float X = sqrtf(-2.0f * logf(R1)) * cosf(2.0f * PI * R2);
	if (X > 4.0f) X = 4.0f;
	if (X < -4.0f) X = -4.0f;
	return X;
}

// Accuracy unknown
inline double atan_fast(double x) {
	return (x / (1.0 + 0.28 * (x * x)));
}


// linear -> dB conversion
inline float lin2dB(float lin) {
	const float LOG_2_DB = 8.6858896380650365530225783783321f;	// 20 / ln( 10 )
	return logf(lin) * LOG_2_DB;
}

// dB -> linear conversion
inline float dB2lin(float dB) {
	const float DB_2_LOG = 0.11512925464970228420089957273422f;	// ln( 10 ) / 20
	return expf(dB * DB_2_LOG);
}

inline bool my_isinf(float f) {
	union {
		float f;
		uint32_t u;
	} f2u;
	f2u.f = f;
	return f2u.u == 0x7f800000 ||
		f2u.u == 0xff800000;
}

inline bool my_isnan(float f) {
	union {
		float f;
		uint32_t u;
	} f2u;
	f2u.f = f;
	// NaNs have non-zero mantissa
	return ((f2u.u & 0x7F800000) == 0x7F800000) && (f2u.u & 0x7FFFFF);
}

inline bool my_isnanorinf(float f) {
	union {
		float f;
		uint32_t u;
	} f2u;
	f2u.f = f;
	// NaNs have non-zero mantissa, infs have zero mantissa. That is, we just ignore the mantissa here.
	return ((f2u.u & 0x7F800000) == 0x7F800000);
}

inline int is_even(float d) {
	float int_part;
	modff(d / 2.0f, &int_part);
	return 2.0f * int_part == d;
}

// Rounds *.5 to closest even number
inline double round_ieee_754(double d) {
	float i = floor(d);
	d -= i;
	if (d < 0.5f)
		return i;
	if (d > 0.5f)
		return i + 1.0f;
	if (is_even(i))
		return i;
	return i + 1.0f;
}

union FP32 {
	uint32_t u;
	float f;
};

struct FP16 {
	uint16_t u;
};

// magic code from ryg: http://fgiesen.wordpress.com/2012/03/28/half-to-float-done-quic/
// See also SSE2 version: https://gist.github.com/rygorous/2144712
inline FP32 half_to_float_fast5(FP16 h)
{
	static const FP32 magic = { (127 + (127 - 15)) << 23 };
	static const FP32 was_infnan = { (127 + 16) << 23 };
	FP32 o;
	o.u = (h.u & 0x7fff) << 13;     // exponent/mantissa bits
	o.f *= magic.f;                 // exponent adjust
	if (o.f >= was_infnan.f)        // make sure Inf/NaN survive
		o.u |= 255 << 23;
	o.u |= (h.u & 0x8000) << 16;    // sign bit
	return o;
}

inline float ExpandHalf(uint16_t half) {
	FP16 fp16;
	fp16.u = half;
	FP32 fp = half_to_float_fast5(fp16);
	return fp.f;
}

// More magic code: https://gist.github.com/rygorous/2156668
inline FP16 float_to_half_fast3(FP32 f)
{
	static const FP32 f32infty = { 255 << 23 };
	static const FP32 f16infty = { 31 << 23 };
	static const FP32 magic = { 15 << 23 };
	static const uint32_t sign_mask = 0x80000000u;
	static const uint32_t round_mask = ~0xfffu;
	FP16 o = { 0 };

	uint32_t sign = f.u & sign_mask;
	f.u ^= sign;

	if (f.u >= f32infty.u) // Inf or NaN (all exponent bits set)
		o.u = (f.u > f32infty.u) ? 0x7e00 : 0x7c00; // NaN->qNaN and Inf->Inf
	else // (De)normalized number or zero
	{
		f.u &= round_mask;
		f.f *= magic.f;
		f.u -= round_mask;
		if (f.u > f16infty.u) f.u = f16infty.u; // Clamp to signed infinity if overflowed

		o.u = f.u >> 13; // Take the bits!
	}

	o.u |= sign >> 16;
	return o;
}

inline uint16_t ShrinkToHalf(float full) {
	FP32 fp32;
	fp32.f = full;
	FP16 fp = float_to_half_fast3(fp32);
	return fp.u;
}

// FPU control.
void EnableFZ();

#endif
