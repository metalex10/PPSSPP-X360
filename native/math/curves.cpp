
#include <math.h>

#include "math/math_util.h"
#include "curves.h"

float linearInOut(int t, int fadeInLength, int solidLength, int fadeOutLength) {
	if (t < 0) return 0;
	if (t < fadeInLength) {
		return (float)t / fadeInLength;
	}
	t -= fadeInLength;
	if (t < solidLength) {
		return 1.0f;
	}
	t -= solidLength;
	if (t < fadeOutLength) {
		return 1.0f - (float)t / fadeOutLength;
	}
	return 0.0f;
}

float linearIn(int t, int fadeInLength) {
	if (t < 0) return 0;
	if (t < fadeInLength) {
		return (float)t / fadeInLength;
	}
	return 1.0f;
}

float linearOut(int t, int fadeOutLength) {
	return 1.0f - linearIn(t, fadeOutLength);
}

float ease(float val) {
	if (val > 1.0f) return 1.0f;
	if (val < 0.0f) return 0.0f;
	return ((-cosf(val * PI)) + 1.0f) * 0.5;
}

float ease(int t, int fadeLength)
{
	if (t < 0) return 0.0f;
	if (t >= fadeLength) return 1.0f;
	return ease((float)t / (float)fadeLength);
}

float sawtooth(int t, int period) {
	return (t % period) * (1.0f / (period - 1));
}

float passWithPause(int t, int fadeInLength, int pauseLength, int fadeOutLength)
{
	if (t < fadeInLength) {
		return -1.0f + (float)t / fadeInLength;
	}
	t -= fadeInLength;
	if (t < pauseLength) {
		return 0.0f;
	}
	t -= pauseLength;
	if (t < fadeOutLength) {
		return (float)t / fadeOutLength;
	}
	return 1.0f;
}
