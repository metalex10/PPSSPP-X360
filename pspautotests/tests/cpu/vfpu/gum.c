/*
FPU Test. Originally from jpcsp project:
http://code.google.com/p/jpcsp/source/browse/trunk/demos/src/fputest/main.c
Modified to perform automated tests.
*/


// For some reason, vectors on the stack all are unaligned. So I've pulled them out
// into globals, which is horribly ugly. But it'll do for now.


#include <common.h>

#include <pspkernel.h>
#include <stdio.h>
#include <string.h>
#include <math.h>

#include <GL/gl.h>
#include <GL/glu.h>
#include <GL/glut.h>

#include <pspgu.h>
#include <pspgum.h>

#include "vfpu_common.h"

ALIGN16 ScePspFVector4 v0, v1, v2;
ALIGN16 ScePspFVector4 matrix[4];

void checkGum(float angle) {
	static ALIGN16 ScePspFMatrix4 m;
	int val = 45;
	float c = cosf(angle);
	float s = sinf(angle);
	static ALIGN16 ScePspFVector3 pos = { 0, 0, -2.5f };
	static ALIGN16 ScePspFVector3 rot;
	rot.x = val * 0.79f * (GU_PI/180.0f);
	rot.y = val * 0.98f * (GU_PI/180.0f);
	rot.z = val * 1.32f * (GU_PI/180.0f);

	sceGuInit();

	sceGumMatrixMode(GU_PROJECTION);
	sceGumLoadIdentity();
	sceGumPerspective(75.0f,16.0f/9.0f,0.5f,1000.0f);
	sceGumStoreMatrix(&m);
	printMatrix("proj", &m);

	sceGumMatrixMode(GU_VIEW);
	sceGumLoadIdentity();
	sceGumStoreMatrix(&m);
	printMatrix("view", &m);

	// Complete Rotation
	sceGumMatrixMode(GU_MODEL);
	sceGumLoadIdentity();
	sceGumTranslate(&pos);
	sceGumRotateXYZ(&rot);
	printf("Vec: (x=%f, y=%f, z=%f)\n", rot.x, rot.y, rot.z);
	printf("Cos/Sin (%f, %f)\n", c, s);
	sceGumStoreMatrix(&m);
	printMatrix("model1", &m);

	// Simple Rotation
	sceGumMatrixMode(GU_MODEL);
	sceGumLoadIdentity();
	sceGumRotateX(30.0f);
	sceGumStoreMatrix(&m);
	printMatrix("model2", &m);
}

void _checkGlRotate() {
	int n;
	float M[16]; 
	
	glMatrixMode(GL_MODELVIEW);
	glLoadIdentity();
	glRotatef(180, 0, 0, 1.0);
	
	glGetFloatv(GL_MODELVIEW_MATRIX, M);
	for (n = 0; n < 4; n++) {
		printf("%f, %f, %f, %f\n", M[n * 4 + 0], M[n * 4 + 1], M[n * 4 + 2], M[n * 4 + 3]);
	}
}

void checkGlRotate(int argc, char *argv[]) {
	glutInit(&argc, argv);
	glutInitDisplayMode(GLUT_RGB | GLUT_DOUBLE | GLUT_DEPTH);
	glutInitWindowSize(480, 272);
	glutCreateWindow(__FILE__);
	_checkGlRotate();
}

int main(int argc, char *argv[]) {
	printf("Started\n");

	resetAllMatrices();
	
	printf("checkGlRotate:\n"); checkGlRotate(argc, argv);
	printf("checkGum 0.7:\n"); checkGum(0.7f);
	printf("checkGum 1.1:\n"); checkGum(1.1f);

	printf("Ended\n");
	return 0;
}
