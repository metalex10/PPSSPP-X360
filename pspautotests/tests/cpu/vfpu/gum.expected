Started
checkGlRotate:
-1.000000, 0.000767, 0.000000, 0.000000
-0.000767, -1.000000, 0.000000, 0.000000
0.000000, 0.000000, 1.000000, 0.000000
0.000000, 0.000000, 0.000000, 1.000000
checkGum 0.7:
proj: 0.733064,0.000000,0.000000,0.000000
proj: 0.000000,1.303225,0.000000,0.000000
proj: 0.000000,0.000000,-1.001001,-1.000000
proj: 0.000000,0.000000,-1.000500,0.000000
view: 1.000000,0.000000,0.000000,0.000000
view: 0.000000,1.000000,0.000000,0.000000
view: 0.000000,0.000000,1.000000,0.000000
view: 0.000000,0.000000,0.000000,1.000000
Vec: (x=0.620465, y=0.769690, z=1.036726)
Cos/Sin (0.764842, 0.644218)
model1: 0.365556,0.906272,0.212227,0.000000
model1: -0.618122,0.065893,0.783316,0.000000
model1: 0.695913,-0.417528,0.584274,0.000000
model1: 0.000000,0.000000,-2.500000,1.000000
model2: 1.000000,0.000000,0.000000,0.000000
model2: 0.000000,0.154251,-0.988032,0.000000
model2: 0.000000,0.988032,0.154251,0.000000
model2: 0.000000,0.000000,0.000000,1.000000
checkGum 1.1:
proj: 0.733064,0.000000,0.000000,0.000000
proj: 0.000000,1.303225,0.000000,0.000000
proj: 0.000000,0.000000,-1.001001,-1.000000
proj: 0.000000,0.000000,-1.000500,0.000000
view: 1.000000,0.000000,0.000000,0.000000
view: 0.000000,1.000000,0.000000,0.000000
view: 0.000000,0.000000,1.000000,0.000000
view: 0.000000,0.000000,0.000000,1.000000
Vec: (x=0.620465, y=0.769690, z=1.036726)
Cos/Sin (0.453596, 0.891207)
model1: 0.365556,0.906272,0.212227,0.000000
model1: -0.618122,0.065893,0.783316,0.000000
model1: 0.695913,-0.417528,0.584274,0.000000
model1: 0.000000,0.000000,-2.500000,1.000000
model2: 1.000000,0.000000,0.000000,0.000000
model2: 0.000000,0.154251,-0.988032,0.000000
model2: 0.000000,0.988032,0.154251,0.000000
model2: 0.000000,0.000000,0.000000,1.000000
Ended
